### DISCUSSION TITLE
simracing backend

### DISCUSSION THREAD

**Discussion by durandom** [theOehrly/Fast-F1#485] [Ideas]
I'm developing a sim-racing telemetry collection pipeline - https://github.com/b4mad/racing

Would it be possible to develop a new backend that pulls data from other sources, that are more related to simracing?
I thought replicating the ergast API would be a fully transparent way to go, but I'm not sure. Looks like the base URL for ergast is hardcoded and not all data acquisition methods are supported by all backends.

Any guidance on how to best approach this? Would you be open to add backends for simracing titles?

**Comment by Casper-Guo** [comment]
Developing sim racing related stuff in this repo sounds very out of scope to me

**Comment by durandom** [comment]
Why? It's the same data structure, just from simulated races. I don't see any F1 specifics in this, except for the name and maybe the event titles.

**Reply by Casper-Guo** [reply]
I meant if we start developing for sim racing in this codebase then there's no argument against developing for nascar, WEC etc. Very soon you've got yourself a bloated repo

**Reply by theOehrly** [reply]
I suppose it's debatable whether it's a good idea to support other racing series (or sim racing) in FastF1. But apart from that, at the moment, there are two major hurdles anyway. First, FastF1 isn't structured in a way that loading data from alternative sources (like a sim racing backend) is possible. Second, I don't have the time to maintain additional backends.

The first problem might be solved in the future with some (major) refactoring of the internal stuff. This is planned anyway to clean up the code, so that there is better separation of concerns with a clear distinction between data representation/manipulation and data gathering/preprocessing. This would make it much easier to add some way for registering alternative backends. Maybe this could be done through external modules that build on top of FastF1 or just register themselves as data provider. So if I can do the refactoring as clean as I would like to, it might become possible for other people to create such an alternative data backend.

I have had the idea of adding support for sim racing data myself, previously. So I did some research and thought about potential features and advantages. In the end, I wasn't really convinced of the idea anymore, mainly for two reasons.
First, comparisons between F1 telemetry and sim racing telemetry aren't easily possible because the sim racing isn't close enough to real world F1 without additional modding.  I think one of the ex F1 engineer youtubers did a video on that some time ago. 
Second, the main selling point of FastF1 is to make the difficult to access F1 telemetry accessible. In my opinion, the data backend is the most important part here. The data representation side of things is not really all that special. So I don't really see a big advantage in reusing that part of FastF1 for other data. Especially given that there will still be problems to make the data compatible.

Regarding your idea of replicating the Ergast API, this will not do. Only a small part of FastF1's data is pulled from Ergast, mainly the results. No timing or telemetry data.

TLDR, it might be possible in the future that other people can add alternative backends. But I don't really see the advantage in doing that.

**Comment by durandom** [comment]
Thanks both for your replies. I might take a stab at refactoring the backends out to their own modules if my schedule permits.

**Reply by theOehrly** [reply]
Sure, feel free to take a look at it. I used the term "refactor" fairly liberaly though. What I meant was "refactor and potentially rewrite parts of the backend". 
Two things need to be achieved. Get rid of data (pre)processing code in classes that are mainly intended for data representation and access. And clean up the whole data processing code which will require to partially rewrite it.
Those two tasks maybe can be done separately. But maybe not.
